File: LinkedList/Dolby/dolby.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
    struct Node* prev;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    newNode->prev = NULL;
    return newNode;
}

struct Node* insertAtBeginning(struct Node* head, int data) {
    struct Node* newNode = createNode(data);

    if (head != NULL) {
        newNode->next = head;
        head->prev = newNode;
    }

    return newNode;
}

struct Node* insertAtEnd(struct Node* head, int data) {
    struct Node* newNode = createNode(data);

    if (head == NULL) {
        return newNode;
    }

    struct Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }

    temp->next = newNode;
    newNode->prev = temp;

    return head;
}

struct Node* insertAfter(struct Node* head, struct Node* prevNode, int data) {
    if (prevNode == NULL) {
        return head;
    }

    struct Node* newNode = createNode(data);
    newNode->prev = prevNode;
    newNode->next = prevNode->next;

    if (prevNode->next != NULL) {
        prevNode->next->prev = newNode;
    }

    prevNode->next = newNode;

    return head;
}

struct Node* insertBefore(struct Node* head, struct Node* nextNode, int data) {
    if (nextNode == NULL) {
        return head;
    }

    struct Node* newNode = createNode(data);
    newNode->prev = nextNode->prev;
    newNode->next = nextNode;

    if (nextNode->prev != NULL) {
        nextNode->prev->next = newNode;
    }

    nextNode->prev = newNode;

    if (nextNode == head) {
        head = newNode;
    }

    return head;
}

struct Node* deleteNode(struct Node* head, struct Node* delNode) {
    if (head == NULL || delNode == NULL) {
        return head;
    }

    if (head == delNode) {
        head = delNode->next;
        if (head != NULL) {
            head->prev = NULL;
        }
    } else {
        if (delNode->prev != NULL) {
            delNode->prev->next = delNode->next;
        }

        if (delNode->next != NULL) {
            delNode->next->prev = delNode->prev;
        }
    }

    free(delNode);

    return head;
}

void printForward(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

void printBackward(struct Node* tail) {
    struct Node* temp = tail;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->prev;
    }
    printf("NULL\n");
}

int main() {
    struct Node* head = NULL;

    head = insertAtEnd(head, 10);
    head = insertAtEnd(head, 20);
    head = insertAtBeginning(head, 5);
    head = insertAfter(head, head->next, 15);
    head = insertBefore(head, head->next, 8); // Inserting before the second node

    printf("Doubly Linked List (Forward): ");
    printForward(head);

    printf("Doubly Linked List (Backward): ");
    // To print backward, find the last node
    struct Node* tail = head;
    while (tail->next != NULL) {
        tail = tail->next;
    }
    printBackward(tail);

    head = deleteNode(head, head->next); // Deleting the second node

    printf("Doubly Linked List after deleting a node: ");
    printForward(head);

    // Free memory
    while (head != NULL) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
    }

    return 0;
}

File: LinkedList/Dolby/Dolby.java

class Node {
    int data;
    Node prev;
    Node next;
    
    public Node(int data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

public class dolby {
    Node head;
    
    public dolby() {
        this.head = null;
    }
    
    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        
        if (head != null) {
            head.prev = newNode;
        }
        
        head = newNode;
    }
    
    public void insertAtEnd(int data) {
        Node newNode = new Node(data);
        
        if (head == null) {
            head = newNode;
            return;
        }
        
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        
        temp.next = newNode;
        newNode.prev = temp;
    }
    
    public void insertAfter(Node prevNode, int data) {
        if (prevNode == null) {
            return;
        }
        
        Node newNode = new Node(data);
        newNode.prev = prevNode;
        newNode.next = prevNode.next;
        
        if (prevNode.next != null) {
            prevNode.next.prev = newNode;
        }
        
        prevNode.next = newNode;
    }
    
    public void deleteNode(Node delNode) {
        if (head == null || delNode == null) {
            return;
        }
        
        if (head == delNode) {
            head = delNode.next;
        }
        
        if (delNode.next != null) {
            delNode.next.prev = delNode.prev;
        }
        
        if (delNode.prev != null) {
            delNode.prev.next = delNode.next;
        }
    }
    
    public void printForward() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }
    
    public void printBackward() {
        Node temp = head;
        while (temp != null && temp.next != null) {
            temp = temp.next;
        }
        
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.prev;
        }
        System.out.println("null");
    }
    
    public static void main(String[] args) {
        dolby list = new dolby();

        list.insertAtEnd(10);
        list.insertAtEnd(20);
        list.insertAtBeginning(5);
        list.insertAfter(list.head.next, 15);

        System.out.print("Doubly Linked List (Forward): ");
        list.printForward();

        System.out.print("Doubly Linked List (Backward): ");
        list.printBackward();

        list.deleteNode(list.head.next); // Deleting the second node

        System.out.print("Doubly Linked List after deleting a node: ");
        list.printForward();
    }
}
